src/compiler/csharp_generator.cc

```diff
@@ -110,6 +110,13 @@ bool GenerateDocCommentBody(grpc::protobuf::io::Printer* printer,
   return GenerateDocCommentBodyImpl(printer, location);
 }
 
+void GenerateObsoleteAttributeIfObsolete(grpc::protobuf::io::Printer* printer,
+                                         const MethodDescriptor* method) {
+  if (method->options().deprecated()) {
+    printer->Print("[global::System.Obsolete]\n");
```

Comment:
I remember there was a problem with the .NET Obsolete attribute that it will break existing users' builds unless it is silenced, so adding the Obsolete attribute to the generated code is as a matter of fact a breaking change (and we try to avoid breaking API changes as much as possible)
So while adding [Obsolete] atrribute for "deprecated" elements may seem to make sense, I not clear to me whether this is actually the right thing to do.

Reply:
Hi Jan,

thanks for you comment! When a method, which is annotated with the obsolete attribute is referenced, a compiler warning is generated. That is why I have disabled the warning in the generated code. We should should not generate warnings without an actual usage. When the project is configured to treat warnings as errors the build fails. So you are right. This is a breaking change. On the other hand, when the server removes the deprecated method eventually the build is broken as well.

What do you think about an top level option in the proto file to generate the C# obsolete attribute? Or some other option which could be used on the command line. This could give a bit more control.

BTW are there any tests which verify that a given proto produces a given C# file?

Here is an example. Given this proto snippet:

```protobuf
rpc SayHelloOld (HelloRequest) returns (HelloReply) {
  option deprecated = true;
}
```
Will produce this C# snippets:

```csharp
[global::System.CodeDom.Compiler.GeneratedCode("grpc_csharp_plugin", null)]
[global::System.Obsolete]
public virtual global::System.Threading.Tasks.Task<global::Helloworld.HelloReply> SayHelloOld(global::Helloworld.HelloRequest request, grpc::ServerCallContext context)
{
  throw new grpc::RpcException(new grpc::Status(grpc::StatusCode.Unimplemented, ""));
}
```
```csharp
[global::System.CodeDom.Compiler.GeneratedCode("grpc_csharp_plugin", null)]
public static grpc::ServerServiceDefinition BindService(GreeterBase serviceImpl)
{
  return grpc::ServerServiceDefinition.CreateBuilder()
    #pragma warning disable CS0612
    .AddMethod(__Method_SayHelloOld, serviceImpl.SayHelloOld)
    #pragma warning restore CS0612
    .AddMethod(__Method_SayHelloNew, serviceImpl.SayHelloNew).Build();
}
```
The complete example is here.
[sample.tar.gz](https://github.com/grpc/grpc/files/9789132/sample.tar.gz)

Best regards,
Chris

Reply:
Hi Jan.

Thanks for your reply! 

It looks like the deprecation of rpc methods is in some grey area. The proto3 language guide contains only an option to deprecate field options. It recommends that languages should create compiler warnings when such a field is used. For the deprecation of messages I can only find some older [issues](https://github.com/protocolbuffers/protobuf/issues/1734) discussing it.

I tested it a bit with different languages and the most languages do not support it. Java and Golang do. C++, C#, Ruby and Python do not.

The documentation and implementation of the deprecation option is inadequate. This leads to confusion because Java devs use it and are wondering why C# devs ignore it.

I like your approach of versioning APIs using namespaces. Maybe because of the existence of the deprecation option it is done in a different way.

Whatever approach we choose it should be consistent. The deprecation option is not and it will be discussed again and again. Either we implement it from end to end or we get rid of it. But that would be not backward compatible.